overlord/snapstate/component.go

```diff
@@ -209,17 +247,24 @@ func doInstallComponent(st *state.State, snapst *SnapState, compSetup *Component
 		discardComp := st.NewTask("discard-component", fmt.Sprintf(i18n.G(
 			"Discard previous revision for component %q"),
 			compSi.Component))
+		componentTS.linkToHook = append(componentTS.linkToHook, discardComp)
 		addTask(discardComp)
 	}
 
-	installSet := state.NewTaskSet(tasks...)
-	installSet.MarkEdge(prepare, BeginEdge)
-	installSet.MarkEdge(linkSnap, MaybeRebootEdge)
+	var postOpHook *state.Task
+	if !compInstalled {
+		postOpHook = SetupInstallComponentHook(st, snapsup.InstanceName(), compSi.Component.ComponentName)
+	} else {
+		postOpHook = SetupPostRefreshComponentHook(st, snapsup.InstanceName(), compSi.Component.ComponentName)
+	}
```

Comment:
I just noticed, `install` and `post-refresh` hooks should happen before the `prepare-kernel-modules-components` task, as they could generate kernel modules that the mentioned task needs to be aware of. Also, `discard-component` should be the last task as it is an undoable task,